openconnect-gp-okta

```diff
@@ -39,22 +45,28 @@ def okta_auth(s, domain, username, password):
             if factor['factorType'] == 'push':
                 url = factor['_links']['verify']['href']
                 while True:
-                    r = post_json(s, url, {'stateToken': r['stateToken']})
+                    r = post_json(s, url, {'factorId': factor['id'], 'stateToken': r['stateToken']})
                     if r['status'] != 'MFA_CHALLENGE':
                         break
                     assert r['factorResult'] == 'WAITING'
                 break
+            if factor['factorType'] == 'sms':
+                url = factor['_links']['verify']['href']
+                r = post_json(s, url, {'factorId': factor['id'], 'stateToken': r['stateToken']})
+                code = input('SMS code? ')
+                r = post_json(s, url, {'factorId': factor['id'], 'stateToken': r['stateToken'], 'passCode': code})
+                break
         else:
             assert False
 
     assert r['status'] == 'SUCCESS'
     return r['sessionToken']
 
-def okta_saml(s, saml_req_url, username, password):
+def okta_saml(s, saml_req_url, saml_req_data, username, password):
     domain = urllib.parse.urlparse(saml_req_url).netloc
 
     # Just to set DT cookie
-    check(s.get(saml_req_url))
+    check(s.request(('POST' if saml_req_data else 'GET'), saml_req_url, data=saml_req_data))
 
     token = okta_auth(s, domain, username, password)
 
```

Comment:
redirectUrl below now doesn't include saml_req_data -- does this work?

Reply:
That's a really good question. I believe the answer is yes, because this data should be encoded in the original `saml_req_url`… but I can't find a live-in-the-wild GP VPN with the redirect method to test on.

So it's probably premature to include this :-1: